### ISSUE TITLE
DPO Loss

### ISSUE THREAD

**Issue by paulcx** [OpenLLMAI/OpenRLHF#235]
I have a question, in the calculation of DPO's loss, should the labels for the prompt portion in chosen_token and reject_token be set to -100, meaning that they are not involved in the calculation of the loss? This method of loss calculation is usually used in the sft training process.

**Comment by hijkzzz** [comment]
the official code DPO does not seem to do so

**Comment by atebbifakhr** [comment]
Actually it does: https://github.com/eric-mitchell/direct-preference-optimization/blob/main/trainers.py#L90-L115

**Comment by paulcx** [comment]
The ｀｀｀_get_batch_logps｀｀｀ function calculates the log probabilities of the non-masked tokens, optionally returning either the sum or the average of these values for each sequence in the batch. The masking mechanism allows for the exclusion of certain tokens (indicated by -100) from the probability calculation, but the function itself does not distinguish between different parts of the input sequence such as prompts and answers. If to do so, we could calculate labels in datasets and set ignore token -100 for each prompt in chosen token and reject token.

**Comment by hijkzzz** [comment]
@atebbifakhr @paulcx 
I observed loss function of DPO, whether mask prompt or not, does not have any effect on the final gradient

![image](https://github.com/OpenLLMAI/OpenRLHF/assets/19810594/1cec162a-5077-4232-ade6-d4d3f8fd2762)

**Comment by paulcx** [comment]
Hi @hijkzzz Would you mind take look at the official code [here](https://github.com/eric-mitchell/direct-preference-optimization/blob/f8b8c0f49dc92a430bae41585f9d467d3618fe2f/preference_datasets.py#L255C1-L277C17)?

**Comment by hijkzzz** [comment]
> Hi @hijkzzz Would you mind take look at the official code [here](https://github.com/eric-mitchell/direct-preference-optimization/blob/f8b8c0f49dc92a430bae41585f9d467d3618fe2f/preference_datasets.py#L255C1-L277C17)?

I see your point, but it seems that the logits in the prompt section will be canceled out by the opposite signs of the chosen and rejected samples due to the presence of the log function.

**Comment by paulcx** [comment]
That's great and I'm willing to test it after that.

**Comment by paulcx** [comment]
Very limited pair testing and the green one is without applying this pr

![image](https://github.com/OpenLLMAI/OpenRLHF/assets/738834/7e61525c-a153-4a64-bf2c-3667d333cbfe)

**Comment by louieworth** [comment]
@paulcx What does the curve represent? Is there any experimental difference between w/ and w/o masks on prompts?

**Comment by louieworth** [comment]
> > Hi @hijkzzz Would you mind take look at the official code [here](https://github.com/eric-mitchell/direct-preference-optimization/blob/f8b8c0f49dc92a430bae41585f9d467d3618fe2f/preference_datasets.py#L255C1-L277C17)?
> 
> I see your point, but it seems that the logits in the prompt section will be canceled out by the opposite signs of the chosen and rejected samples due to the presence of the log function. I can add this mask to make the implementation aligned official implementation.

@hijkzzz has this alignment to the official codebase been made? I can submit a merge for the code.

**Comment by paulcx** [comment]
> @paulcx What does the curve represent? Is there any experimental difference between w/ and w/o masks on prompts?

yes

**Comment by louieworth** [comment]
> > @paulcx What does the curve represent? Is there any experimental difference between w/ and w/o masks on prompts?
> 
> yes

"Yes" mean "no effect" or "effect" on the final performance?

**Comment by paulcx** [comment]
> > > @paulcx What does the curve represent? Is there any experimental difference between w/ and w/o masks on prompts?
> 
> > 
> 
> > yes
> 
> 
> 
> "Yes" mean "no effect" or "effect" on the final performance?

https://x.com/corbtt/status/1806336011804484017

**Comment by louieworth** [comment]
Thanks for sharing this tweet.
